### PR TITLE
API test suite

### PR DESCRIPTION
**Pull request by NivRichter**
# Description

Added API test file to test the backend service of the 3 existing frameworks
4 github actions were added:
Three for each of the frameworks and one for linting our tests

### CODE REVIEW

File: .framework/node/backend/package.json [hunk 1]
```diff
@@ -7,6 +7,7 @@
   },
   "scripts": {
     "start": "node ./app.js",
+    "start-test": "export WILCO_ID=0 && export MONGODB_URI=mongodb://localhost/anythink-market && node ./app.js ",
```
Comment (NivRichter) [inline .framework/node/backend/package.json:10]: will remove, just to make things work

Comment (esapir) [inline .framework/node/backend/package.json:10]: maybe anythink-market-test ? And then your tests will first clear the DB, fill it with dummy data and start tests.

Comment (NivRichter) [reply .framework/node/backend/package.json:10]: yep, can do. I actullay prefer to hide the fact that we override wilcoId, better if users want play with it. I'll check how I can hide it using the github action itself

File: .framework/node/backend/lib/event.js [hunk 2]
```diff
@@ -12,6 +12,9 @@ const axios = axiosLib.create({
 });
 
 async function sendEvent(event, metadata) {
+  if (process.env.APP_ENV === 'TEST') {
+    return
+  }
```
Comment (esapir) [inline .framework/node/backend/lib/event.js:17]: Can we not do that? I prefer to use `nock` or something similar and to not modify code with such ifs

File: .framework/node/backend/lib/event.js [hunk 3]
```diff
@@ -12,8 +12,12 @@ const axios = axiosLib.create({
...
 
 async function sendEvent(event, metadata) {
-  const result = await axios.post(`/users/${WILCO_ID}/event`, JSON.stringify({ event, metadata }));
-  return result.data;
+  try {
+    const result = await axios.post(`/users/${WILCO_ID}/event`, JSON.stringify({event, metadata}));
+    return result?.data;
+  } catch (e) {
+    // pass
+  }
```
Comment (stellaanavi) [inline .framework/node/backend/lib/event.js:20]: Why is the try-catch needed now?

File: .github/workflows/test_api_lint.yml [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    runs-on: ubuntu-latest
+    timeout-minutes: 6
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v3
+        with:
+          # Make sure the actual branch is checked out when running on pull requests
+          ref: ${{ github.head_ref }}
+
+      - name: Run Lint
```
Comment (stellaanavi) [inline .github/workflows/test_api_lint.yml:18]: is this action run lint or adding node to the worker?

File: .framework/node/backend/lib/event.js (hunk 3 above)
Comment (NivRichter) [reply .framework/node/backend/lib/event.js:20]: When the wilco_id is not found in the request to the engine, it returns 4XX, and express throws an error

File: .github/workflows/test_api_lint.yml (hunk 4 above)
Comment (stellaanavi) [reply .github/workflows/test_api_lint.yml:18]: Ohh I see lint run in the last cmd

File: .framework/node/backend/lib/event.js (hunk 3 above)
Comment (stellaanavi) [reply .framework/node/backend/lib/event.js:20]: Can we add env variable to the worker with wilco_id to overcome it?

File: .github/workflows/test_api_lint.yml [hunk 5]
```diff
@@ -0,0 +1,28 @@
+name: Lint
+
+on:
+  pull_request:
+
+jobs:
+  lint:
+    name: Pr checks
+    runs-on: ubuntu-latest
```
Comment (esapir) [inline .github/workflows/test_api_lint.yml:9]: don't use latest. Use specific version

File: .github/workflows/test_api_node.yml [hunk 6]
```diff
@@ -0,0 +1,37 @@
+name: Continuous Integration Node
+
+on:
+  pull_request:
+
+jobs:
+  ci-checks:
+    name: Pr checks
+    runs-on: ubuntu-latest
```
Comment (esapir) [inline .github/workflows/test_api_node.yml:9]: same - don't use latest

File: .github/workflows/test_api_python.yml [hunk 7]
```diff
@@ -0,0 +1,52 @@
+name: Continuous Integration Python
+
+on:
+  pull_request:
+
+jobs:
+  ci-checks:
+    name: Pr checks
+    runs-on: ubuntu-latest
```
Comment (esapir) [inline .github/workflows/test_api_python.yml:9]: also here

File: .github/workflows/test_api_rails.yml [hunk 8]
```diff
@@ -0,0 +1,46 @@
+name: Continuous Integration Rails
+
+on:
+  pull_request:
+
+jobs:
+  ci-checks:
+    name: Pr checks
+    runs-on: ubuntu-latest
```
Comment (esapir) [inline .github/workflows/test_api_rails.yml:9]: and here :)

File: readme.md [hunk 9]
```diff
@@ -8,7 +8,11 @@ Please find more info about each part in the relevant Readme file ([frontend](fr
 
 When implementing a new feature or fixing a bug, please create a new pull request against `main` from a feature/bug branch and add `@vanessa-cooper` as reviewer.
 
-## How to run locally?
+## How to run in dev mode?
+
+### Using Codespace
+1.  run `docker-compose up`
+### On your local machine
```
Comment (esapir) [inline readme.md:15]: what's that?

File: .github/workflows/test_api_rails.yml [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+        working-directory: .framework/rails/backend
+
+      - run: WILCO_ID=0 DATABASE_URL=postgresql://postgres:postgres@localhost/anythink-market bin/rails s >& /dev/null &
+        working-directory: .framework/rails/backend
+
+      - run: yarn install
+        working-directory: tests/API/
+
+      - run: yarn test
+        working-directory: tests/API/
```
Comment (esapir) [inline .github/workflows/test_api_rails.yml:46]: newline in the end of file

**Review by stellaanavi — APPROVED** [review]
Only wilco_id comment, but it is not a blocker in my opinion

File: .github/workflows/test_api_lint.yml (hunk 5 above)
Comment (NivRichter) [reply .github/workflows/test_api_lint.yml:9]: That's a good point, i'll change that. We are using the latest though in the user repo actions

File: readme.md (hunk 9 above)
Comment (NivRichter) [reply readme.md:15]: Well just noticed it along the way. Users that are using codespace don't need to  download docker

File: .github/workflows/test_api_lint.yml (hunk 5 above)
Comment (esapir) [reply .github/workflows/test_api_lint.yml:9]: Not in `node`. But we should change is everywhere

Comment (NivRichter) [reply .github/workflows/test_api_lint.yml]: ah ok

File: .framework/node/backend/lib/event.js (hunk 3 above)
Comment (stellaanavi) [reply .framework/node/backend/lib/event.js:20]: OK got it

Comment (esapir) [inline .framework/node/backend/lib/event.js:20]: @NivRichter Thoughts about it and this code bother me. I think we should not try-catch here but use a Mock Server instead. Maybe something like [mockoon](https://github.com/mockoon/mockoon)?

@shem8 Would be happy to get your input as well

File: .framework/node/backend/lib/event.js [hunk 11]
```diff
@@ -12,8 +12,13 @@ const axios = axiosLib.create({
 });
 
 async function sendEvent(event, metadata) {
-  const result = await axios.post(`/users/${WILCO_ID}/event`, JSON.stringify({ event, metadata }));
-  return result.data;
+  try {
+    const result = await axios.get(`/hello`, JSON.stringify({event, metadata}));
```
Comment (shem8) [inline .framework/node/backend/lib/event.js:16]: why to you call /hello?

File: tests/API/api.test.js [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+      timeout: 10 * 1000,
+    });
+  });
+  it("Server is up and serving requests", async () => {
+    let response;
+    try {
+      response = await healthCheck(client);
+    } catch (error) {
+      if (error.code === "ECONNREFUSED" || !error.response) {
+        console.error("Error: Connection refused by server");
```
Comment (shem8) [inline tests/API/api.test.js:54]: fail

File: tests/API/api.test.js [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+      } else {
+        console.error("Error: Couldn't connect to server", error);
```
Comment (shem8) [inline tests/API/api.test.js:56]: fail

File: tests/API/api.test.js [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+    const username = `user${(Math.random() + 1).toString(36).substring(7)}`;
+    const user = {
+      username: username,
+      email: `${username}@test.work`,
+      password: "wilco1234",
+    };
+    const response = await createUser(client, user);
+    expect(response.data.user.username).toBe(user.username);
+    expect(response.data.user.email).toBe(user.email);
+    expect(response.data.user.token).toBeTruthy();
```
Comment (shem8) [inline tests/API/api.test.js:71]: to not be null

File: tests/API/api.test.js [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+      "Authorization"
+    ] = `Token ${response.data.user.token}`;
+    const itemToCreate = {
+      title: "WilcoTitle",
+      description: "description",
+      tag_list: ["tag1"],
+    };
+    const itemResponse = await createItem(client, itemToCreate);
+
+    const getItem = await client.get(`/api/items/${itemResponse?.slug}`);
```
Comment (shem8) [inline tests/API/api.test.js:92]: expect itemResponse?.slug not to be null

Comment (esapir) [reply tests/API/api.test.js:114]: The tests are here just as a placeholder. @stellaanavi Is going to modify them in any case. This PR is mainly about the infrastructure

File: .framework/node/backend/lib/event.js (hunk 3 above)
Comment (NivRichter) [reply .framework/node/backend/lib/event.js]: Added mock server to handle the events

File: .framework/node/backend/lib/event.js (hunk 11 above)
Comment (NivRichter) [reply .framework/node/backend/lib/event.js]: Was only for test purpses. sorry

File: tests/API/api.test.js (hunk 12 above)
Comment (NivRichter) [reply tests/API/api.test.js:24]: added

File: tests/API/api.test.js (hunk 14 above)
Comment (NivRichter) [reply tests/API/api.test.js]: Adeed

File: tests/API/api.test.js (hunk 13 above)
Comment (NivRichter) [reply tests/API/api.test.js:27]: Added

File: .github/workflows/test_api_node.yml [hunk 16]
```diff
@@ -0,0 +1,43 @@
+name: Continuous Integration Node
```
Comment (esapir) [inline .github/workflows/test_api_node.yml:1]: Let's align the workflow names with the tests folder name. It is either CI or API. Maybe even e2e?

File: tests/API/eventHandler.js [hunk 17]
```diff
@@ -0,0 +1,5 @@
+const eventHandler = (event) => {
+  console.log(event);
+};
+
+module.exports = { eventHandler };
```
[the remark this answers is not in the document]
Comment (NivRichter) [reply tests/API/eventHandler.js:5]: Right, but I'm want to mock only this function and I think i can't mock only partial module

[the remark this answers is not in the document]
Comment (esapir) [reply tests/API/eventHandler.js:5]: mmm.. right, convinced

File: .framework/node/backend/lib/event.js [hunk 18]
```diff
@@ -2,8 +2,8 @@ const axiosLib = require("axios");
 const fs = require("fs");
 
 const WILCO_ID = process.env.WILCO_ID || fs.readFileSync('../.wilco', 'utf8')
+const baseURL = process.env.ENGINE_URL || "https://engine.wilco.gg"
```
[the remark this answers is not in the document]
Comment (NivRichter) [reply .framework/node/backend/lib/event.js:5]: Do we have such env var? I think we replace the engine to the ngrok using logic inside the engine

File: .framework/node/backend/lib/event.js [hunk 19]
```diff
@@ -12,8 +12,8 @@ const axios = axiosLib.create({
 });
 
 async function sendEvent(event, metadata) {
-  const result = await axios.post(`/users/${WILCO_ID}/event`, JSON.stringify({ event, metadata }));
-  return result.data;
+    const result = await axios.post(`/users/${WILCO_ID}/event`, JSON.stringify({event, metadata}));
+    return result?.data;
```
[the remark this answers is not in the document]
Comment (NivRichter) [reply .framework/node/backend/lib/event.js:16]: Nope. removed

File: .github/workflows/test_api_lint.yml [hunk 20]
```diff
@@ -0,0 +1,21 @@
...
+    timeout-minutes: 6
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v3
+        with:
+          # Make sure the actual branch is checked out when running on pull requests
+          ref: ${{ github.head_ref }}
+
+      - run: yarn install
+        working-directory: tests/API
```
[the remark this answers is not in the document]
Comment (NivRichter) [reply .github/workflows/test_api_lint.yml:19]: Correct

File: .framework/node/backend/lib/event.js (hunk 18 above)
[the remark this answers is not in the document]
Comment (esapir) [reply .framework/node/backend/lib/event.js:5]: We do a few things for local environment: we change the URL in code but for actions/codespaces we use env var

File: .github/workflows/test_api_python.yml [hunk 21]
```diff
@@ -0,0 +1,55 @@
...
+          python-version: "3.9.13"
+
+
+      - run: pip install -r requirements.txt
+        working-directory: .framework/python/backend
+
+      - run: pip install gunicorn
+        working-directory: .framework/python/backend
+
+      - run: SECRET_KEY=secret DATABASE_URL=postgresql://postgres:postgres@localhost/anythink-market alembic upgrade head
```
[the remark this answers is not in the document]
Comment (NivRichter) [reply .github/workflows/test_api_python.yml:42]: Good idea, added name

File: .github/workflows/test_api_python.yml [hunk 22]
```diff
@@ -0,0 +1,55 @@
...
+        working-directory: .framework/python/backend
+
+      - run: ENGINE_URL=http://localhost:3003 WILCO_ID=0 SECRET_KEY=secret DATABASE_URL=postgresql://postgres:postgres@localhost/anythink-market  gunicorn app.main:app --worker-class=uvicorn.workers.UvicornWorker --bind=0.0.0.0:3000 --workers=5 >& /dev/null &
+        working-directory: .framework/python/backend
+
+      - run: yarn install
+        working-directory: tests/API/
+
+      - run: yarn wait-on -v http-get://0.0.0.0:3000/health --timeout 20000
```
[the remark this answers is not in the document]
Comment (NivRichter) [reply .github/workflows/test_api_python.yml:51]: reduced to 5

File: tests/API/createUser.js [hunk 23]
```diff
@@ -0,0 +1,34 @@
+const axios = require("axios");
```
[the remark this answers is not in the document]
Comment (NivRichter) [reply tests/API/createUser.js:1]: Deleted

File: tests/API/anytinkClient.js [hunk 24]
```diff
@@ -0,0 +1,49 @@
+const axios = require("axios");
+
+class AnythinkClient {
+  constructor() {
+    this.client = axios.create({
+      baseURL: "http://localhost:3000",
+      timeout: 10 * 1000,
```
[the remark this answers is not in the document]
Comment (NivRichter) [reply tests/API/anytinkClient.js:7]: sure yeah

File: .github/workflows/test_api_node.yml (hunk 16 above)
Comment (NivRichter) [reply .github/workflows/test_api_node.yml:1]: right

File: .framework/node/backend/lib/event.js (hunk 18 above)
[the remark this answers is not in the document]
Comment (NivRichter) [reply .framework/node/backend/lib/event.js]: Thanks changed to `ENGINE_BASE_URL` and added a secrets parsing

File: .github/workflows/test_api_node.yml [hunk 25]
```diff
@@ -0,0 +1,43 @@
...
+jobs:
+  ci-checks:
+    name: Pr checks
+    runs-on: ubuntu-20.04
+    timeout-minutes: 6
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v3
+        with:
+          # Make sure the actual branch is checked out when running on pull requests
```
[the remark this answers is not in the document]
Comment (NivRichter) [reply .github/workflows/test_api_node.yml]: Removed

**Review by esapir — APPROVED** [review]
Looking good

**Comment by NivRichter** [comment]
> Looking good

🚀

**Pull request merged**